From: Michal Kubecek <mkubecek@suse.cz>
Date: Fri, 24 Jun 2016 19:47:00 +0200
Subject: kabi: revert xt_compat_match_from_user signature change
Patch-mainline: Never, kabi workaround
References: CVE-2016-4997 bsc#986362

In 3.12.62, commit 39d710d9a43073d4937468b7b959f8f86640cca5
(netfilter: x_tables: xt_compat_match_from_user doesn't need a
retval), upstream commit 0188346f21e6546498c2a0f84888797ad4063fc5
changes return type of protected function xt_compat_match_from_user().

Dropping that commit would do no harm but it would require unnecessary
adjustments of the following patch; therefore we just revert the
change of return type and put back "return 0" at the end.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 include/linux/netfilter/x_tables.h | 2 +-
 net/netfilter/x_tables.c           | 5 +++--
 2 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/include/linux/netfilter/x_tables.h b/include/linux/netfilter/x_tables.h
index fb4ea1578552..ef93201a7c0e 100644
--- a/include/linux/netfilter/x_tables.h
+++ b/include/linux/netfilter/x_tables.h
@@ -427,7 +427,7 @@ extern void xt_compat_init_offsets(u_int8_t af, unsigned int number);
 extern int xt_compat_calc_jump(u_int8_t af, unsigned int offset);
 
 extern int xt_compat_match_offset(const struct xt_match *match);
-extern void xt_compat_match_from_user(struct xt_entry_match *m,
+extern int xt_compat_match_from_user(struct xt_entry_match *m,
 				     void **dstptr, unsigned int *size);
 extern int xt_compat_match_to_user(const struct xt_entry_match *m,
 				   void __user **dstptr, unsigned int *size);
diff --git a/net/netfilter/x_tables.c b/net/netfilter/x_tables.c
index 45b479329cbf..c4826337866b 100644
--- a/net/netfilter/x_tables.c
+++ b/net/netfilter/x_tables.c
@@ -545,8 +545,8 @@ int xt_compat_match_offset(const struct xt_match *match)
 }
 EXPORT_SYMBOL_GPL(xt_compat_match_offset);
 
-void xt_compat_match_from_user(struct xt_entry_match *m, void **dstptr,
-			       unsigned int *size)
+int xt_compat_match_from_user(struct xt_entry_match *m, void **dstptr,
+			      unsigned int *size)
 {
 	const struct xt_match *match = m->u.kernel.match;
 	struct compat_xt_entry_match *cm = (struct compat_xt_entry_match *)m;
@@ -568,6 +568,7 @@ void xt_compat_match_from_user(struct xt_entry_match *m, void **dstptr,
 
 	*size += off;
 	*dstptr += msize;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(xt_compat_match_from_user);
 
-- 
2.9.0

